server/src/main.rs

```diff
@@ -1,13 +1,130 @@
+use bincode;
 use core::str;
 use rand::Rng;
+use serde::{Deserialize, Serialize};
 use std::{
     io::{Read, Write},
     net::{TcpListener, TcpStream},
+    u32,
 };
 
+#[repr(u32)] // Ensure that the enum is represented as an u32
+#[derive(Serialize, Deserialize)]
+pub enum MessageType {
+    ConnectionRequested = 1,
+    ConnectionAccepted = 2,
+    ConnectionRejected = 3,
+    StubMessage = 4,
+}
+
+impl From<u32> for MessageType {
+    fn from(value: u32) -> MessageType {
+        match value {
+            1 => MessageType::ConnectionRequested,
+            2 => MessageType::ConnectionAccepted,
+            3 => MessageType::ConnectionRejected,
+            4 => MessageType::StubMessage,
+            _ => panic!("Unknown MessageType value: {value}!"),
+        }
+    }
+}
+
+impl Into<u32> for MessageType {
+    fn into(self) -> u32 {
+        self as u32
+    }
+}
+
+pub struct MessageHeader {
+    message_type: u32,
```

Comment:
Message abstraction doesn't have to have length. It's double weird that data and it's length are in a separate structures. We should serialize the message as (MessageType:len:data) and deserialize from this representation but the message abstraction can be just MessageType + Vec<u8>.